ruma-serde/src/canonical_json/value.rs

```diff
@@ -0,0 +1,248 @@
+use std::{
+    collections::BTreeMap,
+    convert::{TryFrom, TryInto},
+    fmt,
+};
+
+use js_int::UInt;
+use serde::de::Error;
+use serde_json::Value as JsonValue;
+
+#[derive(Clone, Eq, PartialEq)]
+pub enum CanonicalJsonValue {
+    /// Represents a JSON null value.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(null).try_into().unwrap();
+    /// ```
+    Null,
+
+    /// Represents a JSON boolean.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(true).try_into().unwrap();
+    /// ```
+    Bool(bool),
+
+    /// Represents a JSON number, whether integer or floating point.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(12).try_into().unwrap();
+    /// ```
+    Number(UInt),
```

Comment:
This should be `Int` actually (canonical JSON doesn't forbit negative values) and the variant name should probably be `Int` / `Integer` too. Also docs talk about floats.